src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  for (let i = 0; i < actions.length; i++) {
+    const action = actions[i];
+
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (let k = 0; k < action.keysToRemove.length; k++) {
+        const key = action.keysToRemove[k];
+
+        delete currentState[key];
+      }
+    }
+
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    states.push(currentState);
```

Comment:
To avoid repeating the action of pushing the currentState into the states array in every case of the switch statement (once you've refactored the if statements into a switch), move this action to the end of the loop after the switch block. This ensures you only write the push operation once, which adheres to the DRY (Don't Repeat Yourself) principle.